src/mlpack/tests/main_tests/krann_test.cpp

```diff
@@ -0,0 +1,335 @@
+#include <string>
+
+#define BINDING_TYPE BINDING_TYPE_TEST
+static const std::string testName = "K-RankApproximateNearestNeighborsSearch";
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
+#include "test_helper.hpp"
+#include <mlpack/methods/rann/krann_main.cpp>
+
+#include <boost/test/unit_test.hpp>
+#include "../test_tools.hpp"
+
+using namespace mlpack;
+
+struct  KRANNTestFixture
+{
+  KRANNTestFixture()
+  {
+    CLI::RestoreSettings(testName);
+  }
+
+  ~KRANNTestFixture()
+  {
+    bindings::tests::CleanMemory();
+    CLI::ClearSettings();
+  }
+};
+
+BOOST_FIXTURE_TEST_SUITE(KRANNMainTest, KRANNTestFixture);
+
+/*
+ * Check that we can't provide reference and query matrices
+ * with different dimensions.
+ */
+BOOST_AUTO_TEST_CASE(KRANNEqualDimensionTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Now we specify an invalid dimension(2) for the query data.
+  // Note that the number of points in queryData and referenceData matrices
+  // are allowed to be different
+  arma::mat queryData;
+  queryData.randu(2, 90); // 90 points in 2 dimensions.
+
+  // Random input, some k <= number of  top tau %ile reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("query", std::move(queryData));
+  SetInputParam("k", (int) 5);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/*
+ * Check that we can't specify an invalid k when only reference
+ * matrix is given.
+ */
+BOOST_AUTO_TEST_CASE(KRANNInvalidKTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k > number of reference points.
+  SetInputParam("reference", referenceData);
+  SetInputParam("k", (int) 101);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["k"].wasPassed = false;
+
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) -1); // Invalid.
+
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  
+  
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["k"].wasPassed = false;
+  
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 6); // Invalid.
+
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
```

Comment:
It might be useful to add a comment to the top of this pointing out that when `SetInputParam()` is called for `reference`, the matrix `referenceData` will be empty (because it was moved earlier).  I think it's good to check the functionality on an empty matrix too though, so we should leave the test. :)